### PR TITLE
Add findApp functionality

### LINKED ISSUE #51: As a user, I can easily filter the appointments by their information.

### PR DESCRIPTION
**Pull request by cliftonchee**
Currently, filter based on NRIC, Date, or exact Time, or any combination . Will continue to work on it. Yet to add test cases.

No frontend changes, will only tell you how many appointments returned.

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -175,6 +175,15 @@ Format: `cancelApp i/NRIC d/DATE from/STARTTIME to/ENDTIME`
...
 
+### Finding appointments: `findApp`
+
+Finds appointments based on the given parameters.
+
+Format: `findApp [i/NRIC] [d/DATE] [from/TIME]`
+
+* Filters an appointment with specific `NRIC`, `DATE` or `TIME`
+* If invalid parameters, error detailing what went wrong will be displayed.
+
```
Comment (limrik) [inline docs/UserGuide.md]: Maybe can provide more details on which appointments are filtered, especially the filter by time, does the command retrieve all appointments starting after the time specified or as long as the time specified falls into the range of the appointment?

Comment (limrik) [reply docs/UserGuide.md]: Also, just curious if providing the time without the date will still fetch me all appointments after that time for all days?

File: src/main/java/seedu/address/logic/commands/FindAppCommand.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Keyword matching is case insensitive.
+ */
+public class FindAppCommand extends Command {
+
+    public static final String COMMAND_WORD = "findApp";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all appointments whose fields contain any of "
+            + "the specified keywords and displays them as a list with index numbers.\n"
+            + "Parameters: KEYWORD [MORE_OPTIONAL_KEYWORDS]...\n"
+            + "Example: " + COMMAND_WORD + " i/ T0123456A d/ 2024-03-16 from/ 11:00";
```
Comment (limrik) [inline src/main/java/seedu/address/logic/commands/FindAppCommand.java]: Similar comment to the one above, "contains any of the specified keywords" seem to imply that "from/ 11" will fetch me all appointments from 11am onwards but we do not actually allow for that as 11 is an invalid time. Not quite sure how we can better phrase this as well, maybe we can discuss further?

File: src/main/java/seedu/address/logic/parser/FindAppCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_NRIC,
+                PREFIX_DATE, PREFIX_START_TIME, PREFIX_END_TIME);
+
+        Optional<Nric> nricFilter = Optional.empty();
+        Optional<Date> dateFilter = Optional.empty();
+        Optional<Time> timeFilter = Optional.empty();
+
+        if (argMultimap.getValue(PREFIX_NRIC).isPresent()) {
+            if (isValidNric(argMultimap.getValue(PREFIX_NRIC).orElse(""))) {
+                nricFilter = Optional.of(new Nric(argMultimap.getValue(PREFIX_NRIC).orElse("")));
```
Comment (limrik) [inline src/main/java/seedu/address/logic/parser/FindAppCommandParser.java]: Could I clarify if there is still this need for the `.orElse("")` with the use of Optional?

File: src/main/java/seedu/address/model/appointment/AppointmentContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import seedu.address.commons.core.date.Date;
+import seedu.address.commons.util.ToStringBuilder;
+import seedu.address.model.person.Nric;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
```
Comment (limrik) [inline src/main/java/seedu/address/model/appointment/AppointmentContainsKeywordsPredicate.java]: Do remember to update JavaDoc comments as well!

File: src/test/java/seedu/address/logic/commands/FindAppCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        FindAppCommand findFirstCommandCopy = new FindAppCommand(firstPredicate);
+        assertTrue(findFirstCommand.equals(findFirstCommandCopy));
+
+        // different types -> returns false
+        assertFalse(findFirstCommand.equals(1));
+
+        // null -> returns false
+        assertFalse(findFirstCommand.equals(null));
+
+        // different person -> returns false
```
Comment (limrik) [inline src/test/java/seedu/address/logic/commands/FindAppCommandTest.java]: Do remember to update comments as well!

File: src/test/java/seedu/address/logic/commands/FindAppCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        String expectedMessage = String.format(MESSAGE_APPOINTMENTS_LISTED_OVERVIEW, 2);
+        AppointmentContainsKeywordsPredicate predicate = preparePredicate(
+                " i/T0123456A from/16:00"
+        );
+        FindAppCommand command = new FindAppCommand(predicate);
+        expectedModel.updateFilteredAppointmentList(predicate);
+        assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertEquals(Arrays.asList(ALICE_APPT, ALICE_APPT_1), model.getFilteredAppointmentList());
+    }
+
```
Comment (limrik) [inline src/test/java/seedu/address/logic/commands/FindAppCommandTest.java]: Perhaps we can add more test cases with invalid fields?

**Review by limrik — COMMENTED** [review]
Thanks Clifton! Just feel that we may have to be clearer in the user guide with the understanding that users may not know what the filter function searches for and see it as a "bug".

File: src/main/java/seedu/address/logic/parser/FindAppCommandParser.java (hunk 3 above)
Comment (cliftonchee) [reply src/main/java/seedu/address/logic/parser/FindAppCommandParser.java]: argMultimap.getValue(PREFIX_NRIC) returns an Optional<String>.

isValidNric requires an argument of type String, hence .orElse(" ") is still required. Hope that clarifies it! 🙏🏻

File: src/test/java/seedu/address/logic/commands/FindAppCommandTest.java (hunk 6 above)
Comment (cliftonchee) [reply src/test/java/seedu/address/logic/commands/FindAppCommandTest.java]: There are some tests in FindAppCommandParserTest.java, would those suffice?

**Review by dayn-1 — APPROVED** [review]
Looks good, thanks @cliftonchee !

**Pull request merged**